server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/SortOperator.java

```diff
@@ -67,75 +108,253 @@ public ListenableFuture<?> isBlocked() {
+  private TsBlock mergeSort() throws IoTDBException {
+
+    if (!diskSpiller.allProcessingTaskFinished()) return null;
+
+    if (mergeSortHeap == null) {
+      mergeSortHeap = new MergeSortHeap(sortReaders.size(), comparator);
+    }
+
+    long startTime = System.nanoTime();
+    long maxRuntime = operatorContext.getMaxRunTime().roundTo(TimeUnit.NANOSECONDS);
+
+    // 1. fill the input from each reader
+    for (int i = 0; i < sortReaders.size(); i++) {
+      if (noMoreData[i] || !isEmpty[i]) continue;
+      SortReader sortReader = sortReaders.get(i);
+      if (sortReader.hasNext()) {
+        MergeSortKey mergeSortKey = sortReader.next();
+        mergeSortKey.tsBlockIndex = i;
+        mergeSortHeap.push(mergeSortKey);
+        isEmpty[i] = false;
+      } else {
+        noMoreData[i] = true;
+        sortBufferManager.releaseOneSortBranch();
+      }
+    }
+
+    // 2. do merge sort until one TsBlock is consumed up
+    tsBlockBuilder.reset();
+    TimeColumnBuilder timeBuilder = tsBlockBuilder.getTimeColumnBuilder();
+    ColumnBuilder[] valueColumnBuilders = tsBlockBuilder.getValueColumnBuilders();
+    while (!mergeSortHeap.isEmpty()) {
+
+      MergeSortKey mergeSortKey = mergeSortHeap.poll();
+      TsBlock targetBlock = mergeSortKey.tsBlock;
+      timeBuilder.writeLong(targetBlock.getTimeByIndex(mergeSortKey.rowIndex));
+      for (int i = 0; i < valueColumnBuilders.length; i++) {
+        if (targetBlock.getColumn(i).isNull(mergeSortKey.rowIndex)) {
+          valueColumnBuilders[i].appendNull();
+          continue;
+        }
+        valueColumnBuilders[i].write(targetBlock.getColumn(i), mergeSortKey.rowIndex);
+      }
+      tsBlockBuilder.declarePosition();
+
+      int readerIndex = mergeSortKey.tsBlockIndex;
+      mergeSortKey = readNextMergeSortKey(readerIndex);
+      if (mergeSortKey != null) {
+        mergeSortHeap.push(mergeSortKey);
+      } else {
+        noMoreData[readerIndex] = true;
+        sortBufferManager.releaseOneSortBranch();
+      }
+
+      // break if time is out or tsBlockBuilder is full or sortBuffer is not enough
+      if (System.nanoTime() - startTime > maxRuntime || tsBlockBuilder.isFull()) {
+        break;
+      }
+    }
     return tsBlockBuilder.build();
   }
 
+  private MergeSortKey readNextMergeSortKey(int readerIndex) throws IoTDBException {
+    SortReader sortReader = sortReaders.get(readerIndex);
+    if (sortReader.hasNext()) {
+      MergeSortKey mergeSortKey = sortReader.next();
+      mergeSortKey.tsBlockIndex = readerIndex;
+      return mergeSortKey;
+    }
+    return null;
+  }
+
+  private boolean hasMoreData() {
+    if (noMoreData == null) return true;
+    for (boolean noMore : noMoreData) {
+      if (!noMore) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public void clear() {
+    if (!diskSpiller.hasSpilledData()) return;
+    try {
+      diskSpiller.clear();
+      Path path = Paths.get(filePrePath);
+      Files.deleteIfExists(path);
```

Comment:
This file won't exist ever. It's a file prefix path, not a directory name.

Reply:
done